### PR TITLE
resetpassword command supports send password reset link and display link

### LINKED ISSUE #29542: [FR] OCC command to trigger a password reset email
Something like `occ user:resetpassword --via-email <uid>` which would use the lost password feature to email a link to the user to reset their account. Account access is blocked until this is used. 

Useful for onboarding / provisioning accounts where you do not want to send the password over email.

### PR DESCRIPTION
**Pull request by sharidas**
The resetpassword supports options to
1) display the password reset link in command line.
2) send email to reset the password for user and
   display the link in the command line.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

The resetpassword supports options to 
1) display the password reset link in command line.
2) send email to reset the password for user and display the link in the command line.

## Related Issue

- Fixes https://github.com/owncloud/core/issues/29542

## Motivation and Context

The resetpassword supports options to 
1) display the password reset link in command line.
2) send email to reset the password for user and display the link in the command line.

## How Has This Been Tested?

- test environment:
- test case 1:
- test case 2:
- ...

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

**Comment by jvillafanez** [comment]
Do we have an use case for both options? Having the "via-mail" option is hard to figure out why we need the "via-link" option.
I'd expect that if the target user doesn't have an email set, you show a warning message in the console (because you couldn't send the email) and show the link (which I'm not really sure if it's needed... how is the link expected to be sent if the admin doesn't know the user email?)

**Comment by jvillafanez** [comment]
In addition, I don't have a clear view of the expected flow.
For example: the command to be focused on setting a new password. The password can be provided interactively or via environment variable. Once the password is correctly set, we send an email to the user to notify the password has been reset.

I don't think this is the flow we want, but the current code doesn't seem to have a clear idea such as the one above.

**Comment by PVince81** [comment]
@jvillafanez the option to output the link on the CLI is intended for docker setups where VMs are created automatically and a custom email is sent by VM scripts, including the link. For example "your server has been setup, click here to set your initial password: XYZ"

**Comment by PVince81** [comment]
As far as I understand this PR is only for generating a password reset token and sending the link by email and/or outputting on the CLI. It is not intended to actually set a password.

**Comment by sharidas** [comment]
Review required @PVince81

File: core/Command/User/ResetPassword.php [hunk 1]
```diff
@@ -60,11 +102,25 @@ protected function configure() {
...
 				'Read the password from the OC_PASS environment variable.'
 			)
+			->addOption(
+				'via-email',
+				null,
+				InputOption::VALUE_NONE,
+				'The email-id set while creating the user, will be used to send link for password reset. This option will also display the link sent to user.'
+			)
+			->addOption(
+				'via-link',
```
Comment (PVince81) [inline core/Command/User/ResetPassword.php]: "via link" means "something will be send through a link" and doesn't reflect what is happening here
maybe call it "output-link" or something similar

and rename "via-email" to "send-email"

File: core/Command/User/ResetPassword.php [hunk 2]
```diff
@@ -74,11 +130,54 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+					'lostpassword', $this->timeFactory->getTime() . ':' . $token);
+
+				$mailData = [
+					'link' => $this->urlgenerator->linkToRouteAbsolute('core.lost.resetform', ['userId' => $userId, 'token' => $token])
+				];
+
+				$mail = new TemplateResponse('core', 'lostpassword/email', $mailData, 'blank');
+				$mailContent = $mail->render();
+
+				$subject = $this->l10n->t('Your password is reset');
```
Comment (PVince81) [inline core/Command/User/ResetPassword.php]: was this copy-pasted from another location ?

File: core/Command/User/ResetPassword.php [hunk 3]
```diff
@@ -74,11 +130,54 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+					return 0;
+				} catch (\Exception $e) {
+					$output->writeln('<error>Can\'t send new user mail to ' . $user->getEMailAddress() . ': ' . $e->getMessage());
+					return 1;
+				}
+			} else {
+				$output->writeln('<error>Email address is not set for the user ' . $user->getUID() . '</error>');
+				return 1;
+			}
+		} elseif ($displayLink) {
```
Comment (PVince81) [inline core/Command/User/ResetPassword.php]: emailing and displaying link aren't mutually exclusive. one could specify both

**Review by PVince81 — CHANGES_REQUESTED** [review]
see comments

File: core/Command/User/ResetPassword.php (hunk 2 above)
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Yes, this was copy-pasted from LostController.

Comment (PVince81) [reply core/Command/User/ResetPassword.php]: can we make the code reusable ? not too happy to see that many lines copy-pasted...

File: core/Command/User/ResetPassword.php (hunk 3 above)
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Updated the code, user can execute either email or displaying link or both.

File: core/Command/User/ResetPassword.php (hunk 1 above)
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Renamed to "send-email" and "output-link".

**Review by PVince81 — CHANGES_REQUESTED** [review]
a few more minor things, see comments

File: core/Command/User/ResetPassword.php (hunk 2 above)
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Agreed. I have refactored the code, in which case at this point of time refactoring LostController, would be too much for this PR, I guess.

File: core/Command/User/ResetPassword.php [hunk 4]
```diff
@@ -74,11 +130,46 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		}
 
 		if ($input->getOption('password-from-env')) {
-			$password = \getenv('OC_PASS');
+			$password = $this->environmentHelper->getEnvVar('OC_PASS');
 			if (!$password) {
 				$output->writeln('<error>--password-from-env given, but OC_PASS is empty!</error>');
 				return 1;
 			}
+		} elseif ($emailLink | $displayLink) {
```
[the remark this answers is not in the document]
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Done

File: core/Command/User/ResetPassword.php [hunk 5]
```diff
@@ -74,11 +130,46 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+				$subject = $this->l10n->t('Your password is reset');
+				try {
+					$message = $this->mailer->createMessage();
+					$message->setTo([$user->getEMailAddress() => $userId]);
+					$message->setSubject($subject);
+					$message->setHtmlBody($mailContent);
+					$message->setFrom([Util::getDefaultEmailAddress('no-reply') => $this->defaults->getName()]);
+					$this->mailer->send($message);
+				} catch (\Exception $e) {
+					$output->writeln('<error>Can\'t send new user mail to ' . $user->getEMailAddress() . ': ' . $e->getMessage());
```
[the remark this answers is not in the document]
Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Added `</error>`.

File: core/Command/User/ResetPassword.php (hunk 2 above)
Comment (PVince81) [reply core/Command/User/ResetPassword.php]: I don't see what has been refactored.

Is there a way then to pick smaller pieces ? I'm a bit worried that for example the token generation + token expiration code is duplicated. If one would decide to change the token format there is a risk of mismatch.

Comment (sharidas) [reply core/Command/User/ResetPassword.php]: Updated the code by modifying LostController, to provide link which can be used by ResetPassword for both send-email and output-link options.

**Review by PVince81 — APPROVED** [review]
Looks better, thanks :+1:

**Comment by sharidas** [comment]
Backport stable10 : https://github.com/owncloud/core/pull/32500

**Pull request merged**